conda_build/metadata.py

```diff
@@ -408,24 +413,32 @@ def _get_all_dependencies(metadata, envs=("host", "build", "run")):
     return reqs
 
 
-def check_circular_dependencies(render_order, config=None):
+def check_circular_dependencies(
+    render_order: list[OutputTuple],
+    config: Config | None = None,
+) -> None:
+    envs: tuple[str, ...]
     if config and config.host_subdir != config.build_subdir:
         # When cross compiling build dependencies are already built
         # and cannot come from the recipe as subpackages
         envs = ("host", "run")
     else:
         envs = ("build", "host", "run")
-    pairs = []
-    for idx, m in enumerate(render_order.values()):
-        for other_m in list(render_order.values())[idx + 1 :]:
+
+    pairs: list[tuple[str, str]] = []
+    for idx, (_, metadata) in enumerate(render_order):
+        name = metadata.name()
+        for _, other_metadata in render_order[idx + 1 :]:
+            other_name = other_metadata.name()
             if any(
-                m.name() == dep or dep.startswith(m.name() + " ")
-                for dep in _get_all_dependencies(other_m, envs=envs)
+                name == MatchSpec(dep).name
```

Comment:
I haven't look at the context here, so this question could not be relevant:
Are we at this point at a state where `dep` is a parsable `MatchSpec` or could it still contain something semi-resolved which isn't a `MatchSpec` quite yet?
If it's always a valid specification, then all good and I'm +1 on removing some of the manual string handling scattered about -- but I just want to make sure it always works.

Reply:
the code was far from clear so to be safe I've reverted the `MatchSpec` change: https://github.com/conda/conda-build/pull/5345/commits/45232e1d17e4608fa7c4bd0c0f16270755662ac0